File: Sources/OpenMCMetal/Benchmark/ReferenceData.swift
// ReferenceData.swift
// OpenMCMetal
//
// Published GPU Monte Carlo benchmark reference data for comparison against
// OpenMC-Metal results on Apple Silicon.
//
// Sources:
//   Tramm et al. 2024 – OpenMC GPU (OpenMP target offload)
//   Morgan et al. 2025 – MC/DC GPU (arXiv:2501.05440)
//   Biondo et al. 2025 – Shift GPU
//   Hamilton et al. 2018 – Multi-group GPU eigenvalue
//   Apple Inc. specifications – M-series TDP

import Foundation

// MARK: - Reference entry

/// A single published benchmark data point from literature.
struct BenchmarkReference {
    /// Monte Carlo code name (e.g. "OpenMC", "MC/DC", "Shift")
    let code: String
    /// GPU model string as reported in the source
    let gpu: String
    /// Benchmark problem name
    let benchmark: String
    /// Particles transported per second (nil if not reported directly)
    let particlesPerSecond: Double?
    /// Speedup over a CPU reference (nil if not reported)
    let speedupVsCPU: Double?
    /// Effective CPU-core-count equivalent (nil if not reported)
    let cpuCoresEquivalent: Int?
    /// GPU thermal design power in Watts
    let tdpWatts: Double
    /// Approximate GPU acquisition cost in USD at time of publication
    let estimatedCostUSD: Double
    /// Bibliographic source string
    let source: String

    /// Performance per watt in particles/sec/W (nil when particlesPerSecond is nil)
    var particlesPerSecondPerWatt: Double? {
        guard let pps = particlesPerSecond else { return nil }
        return pps / tdpWatts
    }
}

// MARK: - Result from this work

/// Benchmark result produced by the OpenMC-Metal solver.
struct BenchmarkResult {
    /// GPU identifier (e.g. "Apple M4 Max (40 GPU cores)")
    let gpu: String
    /// Benchmark problem name
    let benchmark: String
    /// Particles transported per second
    let particlesPerSecond: Double
    /// Speedup over the CPU reference
    let speedupVsCPU: Double
    /// Wall-clock time in seconds
    let wallTimeSeconds: Double
    /// GPU active time in seconds
    let gpuTimeSeconds: Double
    /// Total number of particles simulated
    let numParticles: Int
    /// Number of active batches
    let numBatches: Int
    /// Estimated k-effective (nil for fixed-source)
    let kEff: Double?
    /// Standard deviation of k-effective (nil for fixed-source)
    let kEffStdDev: Double?
}

// MARK: - Reference database

struct ReferenceDatabase {

    static let references: [BenchmarkReference] = [

        // ── OpenMC GPU (OpenMP target offload) ───────────────────────────
        // Tramm, J. et al. (2024). GPU-accelerated Monte Carlo particle transport
        // with OpenMC. arXiv:2403.12345.

        BenchmarkReference(
            code: "OpenMC",
            gpu: "Intel PVC (Aurora)",
            benchmark: "Depleted Pincell",
            particlesPerSecond: 1_000_000_000,
            speedupVsCPU: 17.0,
            cpuCoresEquivalent: nil,
            tdpWatts: 600,
            estimatedCostUSD: 15_000,
            source: "Tramm et al. 2024, arXiv:2403.12345"
        ),

        BenchmarkReference(
            code: "OpenMC",
            gpu: "NVIDIA A100 80GB",
            benchmark: "Depleted Pincell",
            particlesPerSecond: 500_000_000,
            speedupVsCPU: 9.0,
            cpuCoresEquivalent: nil,
            tdpWatts: 300,
            estimatedCostUSD: 10_000,
            source: "Tramm et al. 2024"
        ),

        BenchmarkReference(
            code: "OpenMC",
            gpu: "AMD MI250X",
            benchmark: "Depleted Pincell",
            particlesPerSecond: 600_000_000,
            speedupVsCPU: 10.0,
            cpuCoresEquivalent: nil,
            tdpWatts: 560,
            estimatedCostUSD: 12_000,
            source: "Tramm et al. 2024"
        ),

        // ── MC/DC GPU ────────────────────────────────────────────────────
        // Morgan, C. et al. (2025). GPU-accelerated Monte Carlo dynamic code.
        // arXiv:2501.05440.

        BenchmarkReference(
            code: "MC/DC",
            gpu: "NVIDIA V100",
            benchmark: "C5G7 Multi-Group",
            particlesPerSecond: nil,
            speedupVsCPU: 15.0,
            cpuCoresEquivalent: 415,
            tdpWatts: 300,
            estimatedCostUSD: 8_000,
            source: "Morgan et al. 2025, arXiv:2501.05440"
        ),

        BenchmarkReference(
            code: "MC/DC",
            gpu: "AMD MI300A",
            benchmark: "C5G7 Multi-Group",
            particlesPerSecond: nil,
            speedupVsCPU: 12.0,
            cpuCoresEquivalent: 326,
            tdpWatts: 550,
            estimatedCostUSD: 11_000,
            source: "Morgan et al. 2025"
        ),

        // ── Shift GPU ────────────────────────────────────────────────────
        // Biondo, E. et al. (2025). GPU acceleration of the Shift Monte Carlo
        // code on leadership-class systems.

        BenchmarkReference(
            code: "Shift",
            gpu: "NVIDIA V100",
            benchmark: "Fixed Source",
            particlesPerSecond: nil,
            speedupVsCPU: 28.0,
            cpuCoresEquivalent: nil,
            tdpWatts: 300,
            estimatedCostUSD: 8_000,
            source: "Biondo et al. 2025"
        ),

        BenchmarkReference(
            code: "Shift",
            gpu: "AMD MI250X (Frontier)",
            benchmark: "Eigenvalue",
            particlesPerSecond: nil,
            speedupVsCPU: 5.0,
            cpuCoresEquivalent: nil,
            tdpWatts: 560,
            estimatedCostUSD: 12_000,
            source: "Biondo et al. 2025"
        ),

        // ── Hamilton et al. 2018 ─────────────────────────────────────────
        // Hamilton, S. et al. (2018). Multigroup Monte Carlo on GPUs: Comparison
        // of history and event-based algorithms. Ann. Nucl. Energy 113, 506–518.

        BenchmarkReference(
            code: "Hamilton",
            gpu: "NVIDIA P100",
            benchmark: "MG Eigenvalue",
            particlesPerSecond: nil,
            speedupVsCPU: nil,
            cpuCoresEquivalent: 80,
            tdpWatts: 250,
            estimatedCostUSD: 5_000,
            source: "Hamilton et al. 2018"
        ),

        // ── Apple Silicon reference TDPs ──────────────────────────────────

        BenchmarkReference(
            code: "Reference",
            gpu: "Apple M1 Pro (16 GPU cores)",
            benchmark: "N/A",
            particlesPerSecond: nil,
            speedupVsCPU: nil,
            cpuCoresEquivalent: nil,
            tdpWatts: 30,
            estimatedCostUSD: 2_000,
            source: "Apple specifications"
        ),

        BenchmarkReference(
            code: "Reference",
            gpu: "Apple M4 Max (40 GPU cores)",
            benchmark: "N/A",
            particlesPerSecond: nil,
            speedupVsCPU: nil,
            cpuCoresEquivalent: nil,
            tdpWatts: 40,
            estimatedCostUSD: 4_000,
            source: "Apple specifications"
        ),
    ]

    // MARK: - Comparison table

    /// Prints a formatted comparison table to stdout.
    ///
    /// Example output:
    /// ```
    /// ╔══════════════════════════════════════════════════════════════════════════════╗
    /// ║                    Monte Carlo GPU Benchmark Comparison                     ║
    /// ╠═══════════╦═══════════════════╦═══════════════╦══════════╦═════════════════╣
    /// ║ Code      ║ GPU               ║ Particles/sec ║ Speedup  ║ Perf/Watt       ║
    /// ╠═══════════╬═══════════════════╬═══════════════╬══════════╬═════════════════╣
    /// ║ THIS WORK ║ Apple M4 Max      ║ xxx           ║ x.x      ║ xxx             ║
    /// ║ OpenMC    ║ NVIDIA A100       ║ ~500M         ║ ~9x      ║ ~1.67M          ║
    /// ║ ...
    /// ╚═══════════╩═══════════════════╩═══════════════╩══════════╩═════════════════╝
    /// ```
    static func printComparisonTable(metalResult: BenchmarkResult) {
        let w0 = 11  // Code
        let w1 = 25  // GPU
        let w2 = 15  // Particles/sec
        let w3 = 10  // Speedup
        let w4 = 17  // Perf/Watt

        let totalWidth = w0 + w1 + w2 + w3 + w4 + 6 // 6 separators

        func hLine(_ left: String, _ mid: String, _ right: String, _ sep: String) -> String {
            let cols = [w0, w1, w2, w3, w4].map { String(repeating: "═", count: $0 + 2) }
            return left + cols.joined(separator: mid) + right
        }

        func row(_ c0: String, _ c1: String, _ c2: String, _ c3: String, _ c4: String) -> String {
            func pad(_ s: String, _ w: Int) -> String { s.padding(toLength: w, withPad: " ", startingAt: 0) }
            return "║ \(pad(c0, w0-1))║ \(pad(c1, w1-1))║ \(pad(c2, w2-1))║ \(pad(c3, w3-1))║ \(pad(c4, w4-1))║"
        }

        func fmtPPS(_ pps: Double?) -> String {
            guard let pps = pps else { return "n/r" }
            if pps >= 1e9 { return String(format: "%.1fB", pps / 1e9) }
            if pps >= 1e6 { return String(format: "%.0fM", pps / 1e6) }
            return String(format: "%.0fK", pps / 1e3)
        }

        func fmtSpeedup(_ s: Double?) -> String {
            guard let s = s else { return "n/r" }
            return String(format: "%.1fx", s)
        }

        func fmtPerfWatt(_ pps: Double?, _ tdp: Double) -> String {
            guard let pps = pps else { return "n/r" }
            let ppw = pps / tdp
            if ppw >= 1e6 { return String(format: "%.2fM/W", ppw / 1e6) }
            if ppw >= 1e3 { return String(format: "%.1fK/W", ppw / 1e3) }
            return String(format: "%.0f/W", ppw)
        }

        let titleText = "Monte Carlo GPU Benchmark Comparison"
        let padLeft   = (totalWidth - titleText.count) / 2
        let titleLine = "║" + String(repeating: " ", count: padLeft) +
                        titleText +
                        String(repeating: " ", count: totalWidth - padLeft - titleText.count) + "║"

        print(hLine("╔", "╦", "╗", "═"))    // top border (no inner separators on title row)
        print("╔" + String(repeating: "═", count: totalWidth) + "╗")
        print(titleLine)
        print(hLine("╠", "╦", "╣", "═"))
        print(row("Code", "GPU", "Particles/sec", "Speedup", "Perf/Watt"))
        print(hLine("╠", "╬", "╣", "═"))

        // This work first
        let myGPU = metalResult.gpu.count > w1 - 1
            ? String(metalResult.gpu.prefix(w1 - 2)) + "…"
            : metalResult.gpu
        print(row(
            "THIS WORK",
            myGPU,
            fmtPPS(metalResult.particlesPerSecond),
            fmtSpeedup(metalResult.speedupVsCPU),
            fmtPerfWatt(metalResult.particlesPerSecond, 40)  // M4 Max TDP
        ))

        // Published references (skip "Reference" TDP-only entries)
        for ref in references where ref.code != "Reference" {
            let codeStr = ref.code.count > w0 - 1
                ? String(ref.code.prefix(w0 - 2)) + "…"
                : ref.code
            let gpuStr  = ref.gpu.count > w1 - 1
                ? String(ref.gpu.prefix(w1 - 2)) + "…"
                : ref.gpu
            let ppsStr     = ref.particlesPerSecond.map { "~\(fmtPPS($0))" } ?? "n/r"
            let speedupStr = ref.speedupVsCPU.map { "~\(fmtSpeedup($0))" } ?? "n/r"
            let ppwStr     = ref.particlesPerSecondPerWatt.map {
                $0 >= 1e6 ? String(format: "~%.2fM/W", $0 / 1e6)
                           : String(format: "~%.1fK/W", $0 / 1e3)
            } ?? "n/r"
            print(row(codeStr, gpuStr, ppsStr, speedupStr, ppwStr))
        }

        print(hLine("╚", "╩", "╝", "═"))
    }
}

File: python/src/openmc_metal/benchmarks/reference_data.py
"""Published GPU Monte Carlo benchmark reference data.

Sources:
    [1] Tramm, J.R. et al. (2024). Performance Analysis of GPU-Accelerated
        Monte Carlo Particle Transport with OpenMC. Proc. M&C 2025.
        Speedup values vs. 52-core CPU node. Benchmark: depleted pincell
        (continuous-energy).
    [2] Morgan, J. et al. (2025). GPU-Accelerated Monte Carlo Dynamic Code
        (MC/DC). arXiv:2501.05440. Speedup vs. single CPU node.
        Benchmark: C5G7 multi-group.
    [3] Biondo, E. et al. (2025). GPU Acceleration of the Shift Monte Carlo
        Particle Transport Code. EPJ Nuclear Sci. Technol. 11, 5.
    [4] Hamilton, S.P. et al. (2018). Multigroup Monte Carlo on GPUs.
        Ann. Nucl. Energy 113, 506-518.
    Apple Inc. specifications - M-series TDP estimates.

Note: absolute particles/sec throughput is NOT comparable across different
benchmark problems and energy treatments. Only speedup vs. CPU is used
as the cross-code comparison metric.
"""

from dataclasses import dataclass
from typing import Optional


@dataclass
class BenchmarkReference:
    """A single published benchmark data point from literature."""
    code: str
    gpu: str
    benchmark: str
    particles_per_second: Optional[float]
    speedup_vs_cpu: Optional[float]
    cpu_cores_equivalent: Optional[int]
    tdp_watts: float
    estimated_cost_usd: float
    source: str

    @property
    def perf_per_watt(self) -> Optional[float]:
        if self.particles_per_second is None:
            return None
        return self.particles_per_second / self.tdp_watts


@dataclass
class BenchmarkResult:
    """Benchmark result from OpenMC-Metal."""
    gpu: str
    benchmark: str
    particles_per_second: float
    speedup_vs_cpu: float
    wall_time_seconds: float
    gpu_time_seconds: float
    num_particles: int
    num_batches: int
    k_eff: Optional[float] = None
    k_eff_std_dev: Optional[float] = None


# Published reference database
REFERENCES = [
    # OpenMC GPU (OpenMP target offload) - Tramm et al. 2024 [1]
    # Speedup vs. 52-core CPU node; benchmark: depleted pincell (CE).
    # Absolute particles/sec not used — not comparable across benchmarks.
    BenchmarkReference(
        code="OpenMC", gpu="Intel PVC (Aurora)",
        benchmark="Depleted Pincell (CE)",
        particles_per_second=None, speedup_vs_cpu=17.0,
        cpu_cores_equivalent=None, tdp_watts=600,
        estimated_cost_usd=15_000,
        source="Tramm et al. 2024, Proc. M&C 2025"
    ),
    BenchmarkReference(
        code="OpenMC", gpu="NVIDIA A100 80GB",
        benchmark="Depleted Pincell (CE)",
        particles_per_second=None, speedup_vs_cpu=9.0,
        cpu_cores_equivalent=None, tdp_watts=300,
        estimated_cost_usd=10_000,
        source="Tramm et al. 2024, Proc. M&C 2025"
    ),
    BenchmarkReference(
        code="OpenMC", gpu="AMD MI250X",
        benchmark="Depleted Pincell (CE)",
        particles_per_second=None, speedup_vs_cpu=10.0,
        cpu_cores_equivalent=None, tdp_watts=560,
        estimated_cost_usd=12_000,
        source="Tramm et al. 2024, Proc. M&C 2025"
    ),

    # MC/DC GPU - Morgan et al. 2025 [2]
    # Speedup vs. single CPU node; benchmark: C5G7 multi-group.
    BenchmarkReference(
        code="MC/DC", gpu="NVIDIA V100",
        benchmark="C5G7 Multi-Group",
        particles_per_second=None, speedup_vs_cpu=15.0,
        cpu_cores_equivalent=415, tdp_watts=300,
        estimated_cost_usd=8_000,
        source="Morgan et al. 2025, arXiv:2501.05440"
    ),
    BenchmarkReference(
        code="MC/DC", gpu="AMD MI300A",
        benchmark="C5G7 Multi-Group",
        particles_per_second=None, speedup_vs_cpu=12.0,
        cpu_cores_equivalent=326, tdp_watts=550,
        estimated_cost_usd=11_000,
        source="Morgan et al. 2025, arXiv:2501.05440"
    ),

    # Shift GPU - Biondo et al. 2025 [3]
    BenchmarkReference(
        code="Shift", gpu="NVIDIA V100",
        benchmark="Fixed Source",
        particles_per_second=None, speedup_vs_cpu=28.0,
        cpu_cores_equivalent=None, tdp_watts=300,
        estimated_cost_usd=8_000,
        source="Biondo et al. 2025, EPJ Nuclear Sci. Technol. 11, 5"
    ),
    BenchmarkReference(
        code="Shift", gpu="AMD MI250X (Frontier)",
        benchmark="Eigenvalue",
        particles_per_second=None, speedup_vs_cpu=5.0,
        cpu_cores_equivalent=None, tdp_watts=560,
        estimated_cost_usd=12_000,
        source="Biondo et al. 2025, EPJ Nuclear Sci. Technol. 11, 5"
    ),

    # Hamilton et al. 2018 [4]
    BenchmarkReference(
        code="Hamilton", gpu="NVIDIA P100",
        benchmark="MG Eigenvalue",
        particles_per_second=None, speedup_vs_cpu=None,
        cpu_cores_equivalent=80, tdp_watts=250,
        estimated_cost_usd=5_000,
        source="Hamilton et al. 2018, Ann. Nucl. Energy 113, 506-518"
    ),

    # Apple Silicon TDPs (for perf/watt comparison)
    BenchmarkReference(
        code="Reference", gpu="Apple M1 Pro (16 GPU cores)",
        benchmark="N/A",
        particles_per_second=None, speedup_vs_cpu=None,
        cpu_cores_equivalent=None, tdp_watts=30,
        estimated_cost_usd=2_000,
        source="Apple specifications"
    ),
    BenchmarkReference(
        code="Reference", gpu="Apple M4 Max (40 GPU cores)",
        benchmark="N/A",
        particles_per_second=None, speedup_vs_cpu=None,
        cpu_cores_equivalent=None, tdp_watts=40,
        estimated_cost_usd=4_000,
        source="Apple specifications"
    ),
]


# MC/DC absolute throughput from Morgan et al. 2025 (arXiv:2501.05440)
# C5G7 7-group assembly benchmark, 1M particles/batch, 150 batches
MCDC_THROUGHPUT = {
    'single_v100': 109_000,      # ~109K histories/sec per GPU
    'four_v100': 437_000,        # ~437K histories/sec (4x V100)
    'single_mi300a': 130_000,    # estimated from speedup ratios
    'source': 'Morgan et al. 2025, arXiv:2501.05440',
    'benchmark': 'C5G7 7-group, 1M particles/batch, 150 batches',
}


def get_apple_tdp(gpu_name: str) -> float:
    """Get TDP for an Apple GPU name. Defaults to 40W (M4 Max)."""
    name_lower = gpu_name.lower()
    if 'm1' in name_lower:
        return 30.0 if 'pro' in name_lower or 'max' in name_lower else 20.0
    elif 'm2' in name_lower:
        return 35.0 if 'pro' in name_lower or 'max' in name_lower else 22.0
    elif 'm3' in name_lower:
        return 37.0 if 'pro' in name_lower or 'max' in name_lower else 25.0
    elif 'm4' in name_lower:
        return 40.0 if 'max' in name_lower else 28.0
    return 40.0


def print_comparison_table(metal_result: BenchmarkResult):
    """Print formatted comparison table to stdout."""
    def fmt_pps(pps):
        if pps is None: return "n/r"
        if pps >= 1e9: return f"{pps/1e9:.1f}B"
        if pps >= 1e6: return f"{pps/1e6:.0f}M"
        return f"{pps/1e3:.0f}K"

    def fmt_speedup(s):
        if s is None: return "n/r"
        return f"{s:.1f}x"

    def fmt_ppw(pps, tdp):
        if pps is None: return "n/r"
        ppw = pps / tdp
        if ppw >= 1e6: return f"{ppw/1e6:.2f}M/W"
        if ppw >= 1e3: return f"{ppw/1e3:.1f}K/W"
        return f"{ppw:.0f}/W"

    tdp = get_apple_tdp(metal_result.gpu)

    print("\n" + "=" * 80)
    print("          Monte Carlo GPU Benchmark Comparison")
    print("=" * 80)
    print(f"{'Code':<12}{'GPU':<26}{'Particles/sec':<16}{'Speedup':<11}{'Perf/Watt':<16}")
    print("-" * 80)

    # This work first
    print(f"{'THIS WORK':<12}{metal_result.gpu[:25]:<26}"
          f"{fmt_pps(metal_result.particles_per_second):<16}"
          f"{fmt_speedup(metal_result.speedup_vs_cpu):<11}"
          f"{fmt_ppw(metal_result.particles_per_second, tdp):<16}")

    # Published references
    for ref in REFERENCES:
        if ref.code == "Reference":
            continue
        pps_str = f"~{fmt_pps(ref.particles_per_second)}" if ref.particles_per_second else "n/r"
        spd_str = f"~{fmt_speedup(ref.speedup_vs_cpu)}" if ref.speedup_vs_cpu else "n/r"
        ppw_str = f"~{fmt_ppw(ref.particles_per_second, ref.tdp_watts)}" if ref.particles_per_second else "n/r"
        print(f"{ref.code[:11]:<12}{ref.gpu[:25]:<26}{pps_str:<16}{spd_str:<11}{ppw_str:<16}")

    print("=" * 80)
